### PR TITLE
Bug in query service causing KeyNotFoundException

### PR DESCRIPTION
**Pull request by arunmk-ms**
## Description
Query service used in QIDO request giving a KeyNotFoundException when some of the study instance uids were deleted while the QIDO request is in progress. 

## Related issues
Addresses [issue #105161](https://microsofthealth.visualstudio.com/Health/_workitems/edit/105161).

## Testing
Manual testing in local environment and added unit test for the edge case

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs [hunk 1]
```diff
@@ -168,11 +169,12 @@ private static bool ReturnNewTagDefaults(int? version)
             else
             {
                 Dictionary<string, StudyResult> map = studyComputedResults.ToDictionary<StudyResult, string>(a => a.StudyInstanceUid, StringComparer.OrdinalIgnoreCase);
-                instanceMetadata = instanceMetadata.Select(x =>
-                {
-                    var ds = new DicomDataset(x);
-                    return ds.AddOrUpdate(map[x.GetSingleValue<string>(DicomTag.StudyInstanceUID)].DicomDataset);
-                });
+                instanceMetadata = instanceMetadata.Where(x =>
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs:172]: Instead of this, which does another loop of the list. can you do below on line 176?

File: src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs [hunk 2]
```diff
@@ -168,11 +169,12 @@ private static bool ReturnNewTagDefaults(int? version)
...
+                    map.ContainsKey(x.GetSingleValue<string>(DicomTag.StudyInstanceUID))).Select(x =>
+                    {
+                        var ds = new DicomDataset(x);
+                        return ds.AddOrUpdate(map[x.GetSingleValue<string>(DicomTag.StudyInstanceUID)].DicomDataset);
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs:176]: var key = x.GetSingleValue<string>(DicomTag.StudyInstanceUID);
if(map.ContainsKey(key))
{
}
else
{
}

File: src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs [hunk 3]
```diff
@@ -168,11 +169,12 @@ private static bool ReturnNewTagDefaults(int? version)
             else
             {
                 Dictionary<string, StudyResult> map = studyComputedResults.ToDictionary<StudyResult, string>(a => a.StudyInstanceUid, StringComparer.OrdinalIgnoreCase);
-                instanceMetadata = instanceMetadata.Select(x =>
-                {
-                    var ds = new DicomDataset(x);
-                    return ds.AddOrUpdate(map[x.GetSingleValue<string>(DicomTag.StudyInstanceUID)].DicomDataset);
-                });
+                instanceMetadata = instanceMetadata.Where(x =>
+                    map.ContainsKey(x.GetSingleValue<string>(DicomTag.StudyInstanceUID))).Select(x =>
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs:173]: can you add a comment on this for easy read

File: src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs [hunk 4]
```diff
@@ -151,11 +151,12 @@ private static bool ReturnNewTagDefaults(int? version)
             else
             {
                 Dictionary<DicomIdentifier, SeriesResult> map = seriesComputedResults.ToDictionary<SeriesResult, DicomIdentifier>(a => new DicomIdentifier(a.StudyInstanceUid, a.SeriesInstanceUid, default));
-                instanceMetadata = instanceMetadata.Select(x =>
-                {
-                    var ds = new DicomDataset(x);
-                    return ds.AddOrUpdate(map[new DicomIdentifier(x.GetSingleValue<string>(DicomTag.StudyInstanceUID), x.GetSingleValue<string>(DicomTag.SeriesInstanceUID), default)].DicomDataset);
-                });
+                instanceMetadata = instanceMetadata.Where(x =>
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs:154]: Not your change, can you materialize all the selects in line 141, 149 and 167

Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs:157]: done.

**Review by smithago — APPROVED** [review]
